lib/bridge/IrcHandler.js

```diff
@@ -705,15 +709,26 @@ IrcHandler.prototype.onPart = Promise.coroutine(function*(req, server, leavingUs
     }
 
     let promises = matrixRooms.map((room) => {
-        req.log.info("Leaving room %s", room.getId());
-        return this.ircBridge.getAppServiceBridge().getIntent(
-            matrixUser.getId()
-        ).leave(room.getId());
+        return this.delayWhileInflight(room, matrixUser).then(() => {
+            req.log.info("Leaving room %s", room.getId());
+            return this.ircBridge.getAppServiceBridge().getIntent(
+                matrixUser.getId()
+            ).leave(room.getId());
+        });
     });
     stats.membership(true, "part");
     yield Promise.all(promises);
 });
 
+IrcHandler.prototype.delayWhileInflight = function(room, matrixUser) {
+    if (this.ircBridge.userHasMessagesInFlight(room, matrixUser)) {
+        return Promise.delay(LEAVE_INFLIGHT_DELAY_MS).then(() => {
+            return this.delayWhileInflight(room, matrixUser);
```

Comment:
It's entirely possible that this loops forever. Might be worth having either a maximum number of attempts or maximum delay time. Something like 30 minutes would be fine.

Only reason is to force an exit condition to avoid having the thing spin forever.

Reply:
How would this loop forever?